org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java

```diff
@@ -340,13 +340,14 @@ public void handleClosed(DidCloseTextDocumentParams params) {
 			return;
 		}
 		try {
-			if (JDTUtils.isDefaultProject(unit) || !JDTUtils.isOnClassPath(unit)) {
+			if (JDTUtils.isDefaultProject(unit) || !JDTUtils.isOnClassPath(unit) || unit.getResource().isDerived()) {
 				new DiagnosticsHandler(connection, unit).clearDiagnostics();
 			}
 			if (unit.equals(sharedASTProvider.getActiveJavaElement())) {
 				sharedASTProvider.disposeAST();
 			}
 			unit.discardWorkingCopy();
+			unit.getResource().refreshLocal(IResource.DEPTH_ZERO, null);
```

Comment:
Why do we need a refresh before closing?

Reply:
That should no longer be needed. IIRC, before derived resources where ignored, the client would mess with those compilation units after being modified server-side, and causing errors that would not be discarded after closing the derived file.